### PR TITLE
Add E2E Passphrase verification method to the protocol

### PR DESCRIPTION
**Pull request by tux3**

### CODE REVIEW

File: protocol.md [hunk 1]
```diff
@@ -107,7 +108,7 @@ The first time the *user* starts a Tanker session on a new device, their identit
 Steps to register a new *device*:
 
 1. The *user* uses one the [Verification Method]s available to them to obtain their encrypted [Verification Key]
-1. *Tanker Core* decrypts the encrypted [Verification Key] using the [User Secret]
+1. *Tanker Core* decrypts the encrypted [Verification Key] using either the [User Secret] for a non-E2E method, or the [User Encryption Key Pair] and the E2E method's secret.
```
Comment (blastrock) [inline protocol.md]:
```suggestion
1. *Tanker Core* decrypts the encrypted [Verification Key] using either the [User Secret] for a non-E2E method, or the E2E method's secret.
```

File: protocol.md [hunk 2]
```diff
@@ -72,7 +72,8 @@ A first `device_creation` block is constructed with the *virtual device*'s publi
...
 
-Finally, both `device_creation` *block*s are pushed to the *Trustchain*, the *virtual device* before the *physical one*, and the [Verification Key] is encrypted with the [User Secret] and uploaded to the *Tanker server* according to the [Verification Method] used by the *user*.
+Finally, both `device_creation` *block*s are pushed to the *Trustchain*, the *virtual device* before the *physical one*, and the [Verification Key] is encrypted and uploaded to the *Tanker server* according to the [Verification Method] used by the *user*.
+If the [Verification Method] is an end-to-end method (e.g. the E2E Passphrase method), the [Verification Key] to be uploaded is encrypted once with the [User Encryption Key Pair] and once with the E2E method's secret value (e.g. the E2E passphrase). Otherwise, for non-E2E methods, it is encrypted only with the [User Secret].
```
Comment (JMounier) [inline protocol.md]: - end-to-end or non-E2E choose one

Comment (JMounier) [reply protocol.md]: IMO `end-to-end (E2E)` is at line 186. Use `end-to-end` before that (here and at line 111)

File: protocol.md [hunk 3]
```diff
@@ -180,7 +181,11 @@ A *virtual device* can not be revoked.
...
+For this purpose, during [user registration](#user_registration) *Tanker Core* creates a *virtual device* and pushes its `device_creation` block. This *device* is not a "physical" one, it has no [Local Encrypted Storage]. Instead, the [Device Signature Key Pair] and the [Device Encryption Key Pair] are serialized as an opaque token, we call it the [Verification Key].
+
+There are two types of verification methods: end-to-end (E2E) methods, and non-E2E methods.
+The difference is that for non-E2E methods, the [Verification Key] is encrypted with the [User Secret] and stored on the *Tanker server*, while for E2E methods it is encrypted with the [User Encryption Key Pair] and the E2E method's secret value (e.g. an E2E passphrase) before being stored on the *Tanker server*.
+The reason for this difference is that the [User Secret] can be accessed by parties others than the end-users themselves, namely the *application*, so it is not strictly end-to-end. End-to-end verification methods offer stricter guarantees and do not allow for any possibility of recovery if the verification method's secret value (e.g. passphrase) is lost.
```
Comment (JMounier) [inline protocol.md]:
```suggestion
The reason for this difference is that the [User Secret] can be accessed by parties others than the end-users themselves, namely the *application*, so it is not strictly end-to-end. End-to-end verification methods offer stricter guarantees but do not allow any possibility of recovery if the verification method's secret value (e.g. passphrase) is lost.
```

File: protocol.md [hunk 4]
```diff
@@ -196,6 +201,16 @@ The *user* provides a passphrase when registering a new [Verification Method]. T
...
 
+### E2E Passphrase
+
+This method is an end-to-end (E2E) verification method.
+
+The *user* provides a passphrase when registering a new [Verification Method]. The passphrase will be hashed by *Tanker Core*, with a unique constant value appended, before sending it to the *Tanker server*.
+The passphrase will be hashed a second time by *Tanker Core*, with a different unique constant, and the resulting hash will be used to encrypt the [Verification Key]. The [verification Key] will also be encrypted with the [User Encryption Key Pair]. Then both ciphertexts are sent to the *Tanker server*.
+The *Tanker server* will store the hashed passphrase alongside the user's encrypted [Verification Key].
+
+When the *user* needs to [register a new device](#device-registration), they will provide their passphrase. *Tanker Core* will hash the passphrase, agan with the first unique constant value appended, before sending it to the *Tanker server*. Then, the *Tanker server* will match this result with the stored passphrase. If they match, the *Tanker server* will return the encrypted [Verification Key].
```
Comment (JMounier) [inline protocol.md]:
```suggestion
When the *user* needs to [register a new device](#device-registration), they will provide their passphrase. *Tanker Core* will hash the passphrase, again with the first unique constant value appended, before sending it to the *Tanker server*. Then, the *Tanker server* will match this result with the stored passphrase. If they match, the *Tanker server* will return the encrypted [Verification Key].
```

Comment (blastrock) [inline protocol.md]:
```suggestion
When the *user* needs to [register a new device](#device-registration), they will provide their passphrase. *Tanker Core* will hash the passphrase, again with the first unique constant value appended, before sending it to the *Tanker server*. Then, the *Tanker server* will match this result with the stored passphrase. If they match, the *Tanker server* will return the encrypted [Verification Key].
```

**Review by blastrock — CHANGES_REQUESTED** [review]

File: protocol.md (hunk 2 above)
Comment (tux3) [reply protocol.md]: Ack, switched to end to end for both

File: protocol.md (hunk 4 above)
Comment (tux3) [reply protocol.md]: :+1:

File: protocol.md (hunk 3 above)
Comment (tux3) [reply protocol.md]: Added the but, but I think the for is correct..?

**Review by blastrock — APPROVED** [review]

**Review by JMounier — APPROVED** [review]

**Pull request merged**